packages/firestore/src/local/indexeddb_index_manager.ts

```diff
@@ -516,6 +833,103 @@ export class IndexedDbIndexManager implements IndexManager {
       )
       .next(() => nextSequenceNumber);
   }
+
+  /**
+   * Applies notIn and != filters by taking the provided ranges and excluding
+   * any values that match the `notInValue` from these ranges. As an example,
+   * '[foo > 2 && foo != 3]` becomes  `[foo > 2 && < 3, foo > 3]`.
+   */
+  private applyNotIn(
+    indexRanges: IDBKeyRange[],
+    notInValues: Uint8Array[] | null
+  ): IDBKeyRange[] {
+    if (!notInValues) {
+      return indexRanges;
+    }
+
+    // The values need to be sorted so that we can return a sorted set of
+    // non-overlapping ranges.
+    notInValues.sort((l, r) => compareByteArrays(l, r));
+
+    const notInRanges: IDBKeyRange[] = [];
+    for (const indexRange of indexRanges) {
+      // Use the existing bounds and interleave the notIn values. This means
+      // that we would split an existing range into multiple ranges that exclude
+      // the values from any notIn filter.
+
+      // The first index range starts with the lower bound and ends at the
```

Comment:
I don't think I understand this well..what if `notInValues[0]` is not covered by the first range? What if it is < indexRange.lower?

Reply:
That was indeed completely broken, and also did not work correctly on Android. I ported the Android fix https://github.com/firebase/firebase-android-sdk/pull/3454 and adopted it to work with the Web port.